compat/src/main/scala-2.11_2.12/scala/util/Using.scala

```diff
@@ -0,0 +1,401 @@
+     * this method should be invoked with an `implicit` modifier before the function
+     * parameter:
+     *
+     * Example:
+     * {{{
+     * val lines = Using.Manager { implicit use =>
+     *   new SafeFileReader("file.txt").lines()
+     * }
+     * }}}
+     *
+     * See the main doc for [[Using `Using`]] for full details of suppression behavior.
+     *
+     * @param op the operation to perform using the manager
+     * @tparam A the return type of the operation
+     * @return a [[Try]] containing an exception if one or more were thrown,
+     *         or the result of the operation if no exceptions were thrown
+     */
+    def apply[A](op: Manager => A): Try[A] = Try { (new Manager).manage(op) }
+
+    private final class Resource[R](resource: R)(implicit releasable: Releasable[R]) {
+      def release(): Unit = releasable.release(resource)
+    }
+  }
+
+  private def preferentiallySuppress(primary: Throwable, secondary: Throwable): Throwable = {
+    def score(t: Throwable): Int = t match {
+      case _: VirtualMachineError                   => 4
+      case _: LinkageError                          => 3
+      case _: InterruptedException | _: ThreadDeath => 2
+      case _: ControlThrowable                      => 0
+      case e if !NonFatal(e)                        => 1 // in case this method gets out of sync with NonFatal
+      case _                                        => -1
+    }
+    // special-case `ControlThrowable`, which incorrectly suppresses exceptions
+    //   before 2.13
+    @inline def suppress(t: Throwable, suppressed: Throwable): Throwable = {
+      if (!t.isInstanceOf[ControlThrowable]) t.addSuppressed(suppressed); t
+    }
+
+    if (score(secondary) > score(primary)) suppress(secondary, primary)
+    else suppress(primary, secondary)
+  }
+
+  /** Performs an operation using a resource, and then releases the resource,
+   * even if the operation throws an exception. This method behaves similarly
+   * to Java's try-with-resources.
+   *
+   * $suppressionBehavior
+   *
+   * @param resource the resource
+   * @param body     the operation to perform with the resource
+   * @tparam R the type of the resource
+   * @tparam A the return type of the operation
+   * @return the result of the operation, if neither the operation nor
+   *         releasing the resource throws
+   */
+  def resource[R, A](resource: R)(body: R => A)(implicit releasable: Releasable[R]): A = {
+    if (resource == null) throw new NullPointerException("null resource")
+
+    var toThrow: Throwable = null
+    try {
+      body(resource)
+    } catch {
+      case t: Throwable =>
+        toThrow = t
+        null.asInstanceOf[A] // compiler doesn't know `finally` will throw
+    } finally {
+      if (toThrow eq null) releasable.release(resource)
+      else {
+        try releasable.release(resource)
+        catch { case other: Throwable => toThrow = preferentiallySuppress(toThrow, other) } finally throw toThrow
```

Comment:
why is this `finally` not on its own line? formatting is hard

Reply:
ha, is scalafmt doing that? that seems.... suboptimal

Reply:
presumably

I'll see if I can faff around with it and get it to not look horrible like that